llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,224 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass converts Loop Iteration Count Metadata to Assumptions which can be
+// picked up by Loop Rotate to remove Loop Guards.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumptions"
+
+using namespace llvm;
+
+namespace {
+
+std::string getFunctionAndBlockNames(BasicBlock &BB) {
+  return BB.getParent()->getName().str() + " " + BB.getName().str();
+}
+
+/// Return the Branch Compare Instruction of CurrentLoop if the Loop is well
+/// formed and this pass can process the Predicate
+ICmpInst *getLoopCmpInst(Loop &CurrentLoop) {
+
+  if (CurrentLoop.isRotatedForm()) {
+    LLVM_DEBUG(dbgs() << "Loop already in rotated form. Will not add Loop "
+                         "Iteration Count assumptions.\n");
+    return nullptr;
+  }
+
+  /// Check that the loop has a single Exiting Block. If the CurrentLoop
+  /// contains multiple Exiting Block, ExitBB will be a nullptr
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return nullptr;
+
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI)
+    return nullptr;
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return nullptr;
+
+  LLVM_DEBUG(dbgs() << "Condition Found: " << *LoopCmpInstr << "\n");
+
+  // Do not process equal and not-equal Compare Instructions, since they do not
+  // work well with the currently used assumption-construction-method of reusing
+  // the ICmpInst Predicate
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return nullptr;
+  }
+  return LoopCmpInstr;
+}
+
+/// Return the AddRecExpr evaluated at Iteration 0 if an AddRecExpr can be
+/// extracted, otherwise return Op
+Value *expandValueAtZeroIteration(Value *Op, ScalarEvolution &SE,
+                                  SCEVExpander &Expander,
+                                  Instruction *InsertionPoint) {
+  const SCEVAddRecExpr *AddRec =
+      dyn_cast_or_null<SCEVAddRecExpr>(SE.getSCEV(Op));
+  if (!AddRec) {
+    LLVM_DEBUG(dbgs() << "Could not extract AddRecExpr, will reuse " << *Op
+                      << "\n");
+    return Op;
```

Comment:
I think this is dangerous: we might return a `LoadInst` that isn't loop-invariant. What we return here must be loop-invariant because we will later clone it.

E.g. we could have
```
loop.header:
  store ...
  %iv = load ...
```

The store location could alias with the load location.

Reply:
So we might want to run LICM beforehand, or use analysis like `MemorySSA` or `LoopAccessAnalysis` to help us say what is loop invariant.

Reply:
Also note `LoopUtils` has a bunch of things to help you. In particular:
```
/// Returns true if is legal to hoist or sink this instruction disregarding the
/// possible introduction of faults.  Reasoning about potential faulting
/// instructions is the responsibility of the caller since it is challenging to
/// do efficiently from within this routine.
/// \p TargetExecutesOncePerLoop is true only when it is guaranteed that the
/// target executes at most once per execution of the loop body.  This is used
/// to assess the legality of duplicating atomic loads.  Generally, this is
/// true when moving out of loop and not true when moving into loops.
/// If \p ORE is set use it to emit optimization remarks.
bool canSinkOrHoistInst(Instruction &I, AAResults *AA, DominatorTree *DT,
                        Loop *CurLoop, MemorySSAUpdater &MSSAU,
                        bool TargetExecutesOncePerLoop,
                        SinkAndHoistLICMFlags &LICMFlags,
                        OptimizationRemarkEmitter *ORE = nullptr);
```

Reply:
Now that we have a way to check that something is loop-invariant, do we really need the hoisting? IMO `expandValueAtZeroIteration` should handle two cases:
 - AddRec SCEVs for which we can give a value at iteration zero
 - And loop-invariant expressions, that have the same value at each iteration

Then if `expandValueAtZeroIteration` returns a non-null value for both `rhs` and `lhs`, we can insert the assumption in the header. That assumption will hold for every iteration, so no need to hoist it.

Reply:
We could get away without hoisting/inserting the assumption in the preheader, when LICM has run and we use the incoming value of the PHINode from the preheader in the assumption. 
I found it nicer, to add the assumption in the preheader to make it clear, that the assumption is loop invariant.